sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -312,105 +315,141 @@ public TableRow apply(TableRow input) {
-    return new AutoValue_BigQueryIO_Read.Builder()
+    return new Read();
+  }
+
+  /**
+   * A {@link PTransform} that reads from a BigQuery table and returns a
+   * {@link PCollection} of records transformed by a user defined conversion function.
+   */
+  public static <T>ReadGenericRecords<T> readRecords(
+      SerializableFunction<SchemaAndRecord, T> parseFn) {
+    return new AutoValue_BigQueryIO_ReadGenericRecords.Builder<T>()
         .setValidate(true)
         .setWithTemplateCompatibility(false)
         .setBigQueryServices(new BigQueryServicesImpl())
+        .setParseFn(parseFn)
         .build();
   }
 
-  /** Implementation of {@link #read}. */
-  @AutoValue
-  public abstract static class Read extends PTransform<PBegin, PCollection<TableRow>> {
-    @Nullable abstract ValueProvider<String> getJsonTableRef();
-    @Nullable abstract ValueProvider<String> getQuery();
-    abstract boolean getValidate();
-    @Nullable abstract Boolean getFlattenResults();
-    @Nullable abstract Boolean getUseLegacySql();
+  @VisibleForTesting
+  static class TableRowParser
+      implements SerializableFunction<SchemaAndRecord, TableRow> {
 
-    abstract Boolean getWithTemplateCompatibility();
+    public static final TableRowParser INSTANCE = new TableRowParser();
 
-    abstract BigQueryServices getBigQueryServices();
-    abstract Builder toBuilder();
+    public TableRow apply(SchemaAndRecord schemaAndRecord) {
+      return BigQueryAvroUtils.convertGenericRecordToTableRow(
+          schemaAndRecord.getRecord(),
+          schemaAndRecord.getTableSchema());
+    }
+  }
 
-    @AutoValue.Builder
-    abstract static class Builder {
-      abstract Builder setJsonTableRef(ValueProvider<String> jsonTableRef);
-      abstract Builder setQuery(ValueProvider<String> query);
-      abstract Builder setValidate(boolean validate);
-      abstract Builder setFlattenResults(Boolean flattenResults);
-      abstract Builder setUseLegacySql(Boolean useLegacySql);
+  /**
+   * Implementation of {@link BigQueryIO#read()}.
+   */
+  public static class Read
+      extends PTransform<PBegin, PCollection<TableRow>> {
+    private final ReadGenericRecords<TableRow> inner;
 
-      abstract Builder setWithTemplateCompatibility(Boolean useTemplateCompatibility);
+    Read() {
+      this(BigQueryIO.readRecords(TableRowParser.INSTANCE).withCoder(TableRowJsonCoder.of()));
+    }
 
-      abstract Builder setBigQueryServices(BigQueryServices bigQueryServices);
-      abstract Read build();
+    Read(BigQueryIO.ReadGenericRecords<TableRow> inner) {
+      this.inner = inner;
     }
 
-    /** Ensures that methods of the from() / fromQuery() family are called at most once. */
-    private void ensureFromNotCalledYet() {
-      checkState(
-          getJsonTableRef() == null && getQuery() == null, "from() or fromQuery() already called");
+    @Override
+    public PCollection<TableRow> expand(PBegin input) {
+      return inner.expand(input);
```

Comment:
Better do input.apply(inner). expand must not be called directly by anybody except a runner. I think you also don't need populateDisplayData, validate, or others - inner will do this for you automatically.

Reply:
fwiw I was just copying what `TextIO.Write` does here.  Also removing the delegation to populateDisplayData, etc breaks some unit tests.  Should I still drop the overrides and change the `expand` to what you suggested?

Reply:
Hm, I guess it doesn't hurt to keep the populateDisplayData method; but you definitely should change .expand() to .apply(). And validate() is definitely unnecessary.

Reply:
👍 sounds good.